video/vdu_buffered.h

```diff
@@ -0,0 +1,176 @@
+#ifndef VDU_BUFFERED_H
+#define VDU_BUFFERED_H
+
+#include <memory>
+#include <vector>
+#include <unordered_map>
+
+#include "agon.h"
+#include "buffer_stream.h"
+#include "multi_buffer_stream.h"
+#include "types.h"
+
+std::unordered_map<uint16_t, std::vector<std::shared_ptr<BufferStream>, psram_allocator<std::shared_ptr<BufferStream>>>> buffers;
+
+// VDU 23, 0, &A0, bufferId; command: Buffered command support
+//
+void VDUStreamProcessor::vdu_sys_buffered() {
+	auto bufferId = readWord_t();
+	auto command = readByte_t();
+
+	switch (command) {
+		case BUFFERED_WRITE: {
+			bufferWrite(bufferId);
+		}	break;
+		case BUFFERED_CALL: {
+			bufferCall(bufferId);
+		}	break;
+		case BUFFERED_CLEAR: {
+			bufferClear(bufferId);
+		}	break;
+		case BUFFERED_CREATE: {
+			bufferCreate(bufferId);
+		}	break;
+		case BUFFERED_SET_OUTPUT: {
+			setOutputStream(bufferId);
+		}	break;
+		case BUFFERED_ADJUST: {
+			bufferAdjust(bufferId);
+		}	break;
+		case BUFFERED_DEBUG_INFO: {
+			debug_log("vdu_sys_buffered: buffer %d, %d streams stored\n\r", bufferId, buffers[bufferId].size());
+			// output contents of buffer stream 0
+			auto buffer = buffers[bufferId][0];
+			auto bufferLength = buffer->size();
+			for (auto i = 0; i < bufferLength; i++) {
+				auto data = buffer->getBuffer()[i];
+				debug_log("%02X ", data);
+			}
+			debug_log("\n\r");
+		}	break;
+		default: {
+			debug_log("vdu_sys_buffered: unknown command %d, buffer %d\n\r", command, bufferId);
+		}	break;
+	}
+}
+
+// VDU 23, 0, &A0, bufferId; 0, length; data...: store stream into buffer
+// This adds a new stream to the given bufferId
+// allowing a single bufferId to store multiple streams of data
+//
+void VDUStreamProcessor::bufferWrite(uint16_t bufferId) {
+	auto length = readWord_t();
+	auto bufferStream = make_shared_psram<BufferStream>(length);
+
+	debug_log("bufferWrite: storing stream into buffer %d, length %d\n\r", bufferId, length);
+
+	for (auto i = 0; i < length; i++) {
+		auto data = readByte_b();
+		bufferStream->writeBufferByte(data, i);
+	}
+
+	buffers[bufferId].push_back(std::move(bufferStream));
+	debug_log("bufferWrite: stored stream in buffer %d, length %d, %d streams stored\n\r", bufferId, length, buffers[bufferId].size());
+}
+
+// VDU 23, 0, &A0, bufferId; 1: Call buffer
+// Processes all commands from the streams stored against the given bufferId
+//
+void VDUStreamProcessor::bufferCall(uint16_t bufferId) {
+	debug_log("bufferCall: buffer %d\n\r", bufferId);
+	if (buffers.find(bufferId) != buffers.end()) {
+		auto streams = buffers[bufferId];
+		auto multiBufferStream = make_shared_psram<MultiBufferStream>(streams);
+		// TODO consider - should this use originalOutputStream?
+		auto streamProcessor = make_unique_psram<VDUStreamProcessor>(multiBufferStream, outputStream);
```

Comment:
the question as to whether or not a `bufferCall` should use the current `outputStream` or the `originalOutputStream` is really about whether commands that send a response packet should by default send their responses to the eZ80 or not, coupled with considerations around the "set output buffer" command.

one possibility here may be to allow VDUStreamProcessor to be created passing in 3 streams - input, output, and originalOutput - that way the "original output" will always be the eZ80.